tests/test_coverage.py

```diff
@@ -0,0 +1,309 @@
+        pass
+
+    # false
+    collector = Collector(f)
+    with collector:
+        f(0)
+    assert collector.branches == {((1, 11), (7, 12))}
+
+    # true
+    collector = Collector(f)
+    with collector:
+        f(1)
+    assert collector.branches == {((1, 11), (2, 15)), ((2, 15), (3, 16))}
+
+
+def test_while():
+    def f():
+        n = 0
+        while n < 5:
+            n += 1
+        pass
+
+    collector = Collector(f)
+    with collector:
+        f()
+    assert collector.branches == {
+        # I'm not sure what this (2, 14) self-reference branch is for. TODO look
+        # at the bytecode here
+        ((2, 14), (2, 14)),
+        ((2, 14), (3, 12)),
+        ((2, 14), (4, 8)),
+    }
+
+
+def test_while_initial_false():
+    def f(x):
+        while x == 1:
+            pass
+        pass
+
+    collector = Collector(f)
+    with collector:
+        f(0)
+    assert collector.branches == {((1, 14), (3, 8))}
+
+
+def test_for_no_else_no_break():
+    # `for` branches from the `for` statement itself to:
+    # * the definition of the loop variable `_`, during a loop iteration
+    # * the `pass` statement after the loop, when the loop ends
+    def f():
+        for _ in range(10):
+            pass
+        pass
+
+    with Collector(f) as collector:
+        f()
+    assert collector.branches == {((1, 8), (3, 8)), ((1, 8), (1, 12))}
+
+
+def test_for_no_else_break():
+    # Ideally the break statement would generate a branch ((2, 12), (3, 8)) here,
+    # but it doesn't. See test_for_loop_break_does_not_branch.
+    def f():
+        for _ in range(10):
+            break
+        pass
+
+    with Collector(f) as collector:
+        f()
+    assert collector.branches == {((1, 8), (1, 12))}
+
+
+def test_for_else_no_break():
+    def f():
+        for _ in range(10):  # L1
+            pass  # L2
+        else:
+            pass  # L4
+        pass
+
+    with Collector(f) as collector:
+        f()
+    assert collector.branches == {((1, 8), (1, 12)), ((1, 8), (4, 12))}
+
+
+def test_for_else_break():
+    def f():
+        for _ in range(10):
+            break
+        else:
+            pass
+        pass
+
+    with Collector(f) as collector:
+        f()
+    assert collector.branches == {((1, 8), (1, 12))}
+
+
+def test_for_loop_break_does_not_branch():
+    # `break` inside of a for loop does not generate a branch event. This is
+    # unfortunate for us, because we can't distinguish a for loop which completes
+    # normally from a for loop which breaks early.
+    def f(*, should_break):
+        for i in range(10):
+            if i == 5 and should_break:
+                break
+        pass
+
+    with Collector(f) as collector:
+        f(should_break=False)
+    branches = collector.branches
+
+    with Collector(f) as collector:
+        f(should_break=False)
+    # this assertion is descriptive, not perscriptive. We wish there was a new
+    # branch here.
+    assert collector.branches == branches
```

Comment:
This was the only case I've found so far where `sys.monitoring` is not ideal.

Overall I've been pretty happy with how closely it matches what I expect. `sys.monitoring` deals in bytecode instructions, not in line/column numbers, so you have to adjust expectations sometimes; you might expect that this code:

```python
if a == 1:
    pass
b = 1
```

with `a = 0` has a branch from line 1 col 0 to line 3 col 0, but it really has a branch from line 1 col 4 (the `a` var) to line 3 col 4 (to LOAD 1 onto the stack, I think). The arcs are strictly internal, so I'm fine with this.

Reply:
I note you've got `should_break=False` twice there.

Also iirc `coverage` deals with this somehow...

Reply:
Thanks for catching that (it didn't change the conclusion, though). I'll look at how coverage does this, but I wouldn't be too surprised to find the answer is a manual bytecode analysis

Reply:
It might be fine that this doesn't count as a branch actually? If `break` executes in one iteration and not another, it must have been because some other branch was taken in one and not the other, and that branch would distinguish this break behavior. 

Exceptions/returns might be nonobservable control flow here though

Reply:
Could get those via CALL / RETURN / {PY_,C_}UNWIND events maybe?  Seems fine to revisit that later though.